### PR TITLE
[ENH]: data kwarg support for mplot3d #20912

### LINKED ISSUE #20912: [ENH]: data kwarg support for mplot3d
### Problem

3d axes don't support the data kwarg:
```
gcf().add_subplot(projection="3d").scatter("a", "b", "c", data={"a": [0], "b": [1], "c": [2]})
```
results in
```
ValueError: could not convert string to float: 'a'
```

### Proposed solution

I *think* it's "mostly" a matter of adding a bunch of `@_preprocess_data` decorators to 3D plotting methods similarly to what's done for 2D plots, hence tagging as good-first-issue (though you'll need to understand a bit about the Matplotlib internals).

### Additional context and prior art

_No response_

### PR DESCRIPTION
**Pull request by jayjoshi112711**
## PR Summary
Fix #20912

Hi, I have started working on this issue. This will be my first open source contribution, so I will need some help and guidance. I have gone through the contributing guide and have setup the development environment and Matplotlib in editable mode on my system. 

I am currently working on understanding the Matplotlib internals.

### CODE REVIEW

**Comment by story645** [comment]
Hi, thanks for the commit but you accidentally included a whole bunch of files. Please remedy that by 

1. stash any current changes https://git-scm.com/docs/git-stash
2. checkout master
3. create a new branch off master
4. add your changes to the new branch
5. just add and commit the files you changed
6. push. It's fine if you open a new PR to do this as I figure starting over with a new branch will be easier than working with this one.
If you need help, please ask on gitter, especially the incubator channel https://matplotlib.org/devdocs/devel/contributing.html#contributor-incubator

**Comment by timhoffm** [comment]
You can force-push over this PR.

1. `git checkout 20912-data-kwarg-support-mplot3d`
2. cleanup your branch
3. `git push origin 20912-data-kwarg-support-mplot3d --force`

There are multiple ways to cleanup a branch, with more or less git magic.
If that's too complicated, a semi-manual way ist:

2.1) Backup the changes you still want.
2.2) `git reset --hard master`
2.3) Apply your changes again.

Anyway, the important message is: you can `git push ... --force` a branch for a PR to replace everything in there.

**Comment by jayjoshi112711** [comment]
Hi, I have added the @_preprocess_data decorator to the scatter method in mplot3d/axes3d.py and it seems to work properly. 

Code: 
import matplotlib.pyplot as plt
>>> plt.gcf().add_subplot(projection="3d").scatter("a", "b", "c", data={"a": [0], "b": [1], "c": [2]})
<mpl_toolkits.mplot3d.art3d.Path3DCollection object at 0x7f80872b9400>
>>> plt.show()
Figure_1

It gives the desired output.

I have a couple of queries. 
1. What labels should be included by default  in replace names and label_namer.
2. How to resolve the checks that are currently failing.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 1]
```diff
@@ -2271,6 +2271,10 @@ def add_collection3d(self, col, zs=0, zdir='z'):
         collection = super().add_collection(col)
         return collection
 
+    @_preprocess_data(replace_names=["xs", "ys", "zs", "s",
+                                     "edgecolors", "c", "facecolor",
+                                     "facecolors", "color"],
+                      label_namer=None)
```
Comment (timhoffm) [inline lib/mpl_toolkits/mplot3d/axes3d.py:2277]:
```suggestion
                                     "facecolors", "color"])
```
It's corrct that there is no canonical label namer in this case, but `None` is the default so can be left out.

**Comment by timhoffm** [comment]
> How to resolve the checks that are currently failing.

Place this before the `**kwargs` parameter entry in the docstring.
```
        data : indexable object, optional
            DATA_PARAMETER_PLACEHOLDER
``` 
We auto-generate the content for the description.

**Comment by jayjoshi112711** [comment]
Hi, 
I have added the @_preprocess_data decorator to relevant methods in mplot3d.py. All the tests are passing except Codecov/project/tests. Is it safe to merge?

**Review by timhoffm — APPROVED** [review]

**Comment by story645** [comment]
Is it ok that this isn't tested?

**Comment by timhoffm** [comment]
I think that's okish not to have tests. `preprocess_data()` itself is tested extensively. So tests would only add integration testing between the decorator and the 3d plotting functions. We also don't explicitly test combininations of positional and keyword arguments.

Systematically adding new tests for data kwarg an all functions seems overkill. Maybe one could add one or two image comparison tests and/or change some existing tests to use `data`. But given this is a first time contribution, I want to keep the barrier low.

**Comment by story645** [comment]
Congrats on your first pr @jayjoshi112711! 🥳
We hope to see you around!

**Comment by jayjoshi112711** [comment]
Thanks, @story645 and @timhoffm for helping me with my first PR, looking forward to working on more issues :)
Cheers!

**Comment by QuLogic** [comment]
@meeseeksdev backport to v3.5.x

**Pull request merged**